### PR TITLE
BUG: Allow `int` to be called on nested object arrays, fix `np.str_.__int__`

### PR DESCRIPTION
**Pull request by eric-wieser**
Rather than assuming that any object array is self-referencing, we can just use PyEnter_RecursiveCall, in:

* `__int__`
* `__float__`
* `__long__`
* `__hex__`
* `__oct__`

This works towards (but does _not_ fix) #9972, by not trying to touch the `nb_*` slots ourselves (the same way as #10040).

Substantial code deduplication here. Error message is different, but perhaps also better:

```python
>>> int(np.array([None]))
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType' # now
TypeError: cannot convert to an int; scalar object is not a number #before
```

### CODE REVIEW

File: numpy/core/src/multiarray/number.c [hunk 1]
```diff
@@ -814,213 +814,107 @@ _array_nonzero(PyArrayObject *mp)
...
-     * If we still got an array which can hold references, stop
-     * because it could point back at 'v'.
-     */
-    if (PyArray_Check(pv) &&
-                    PyDataType_REFCHK(PyArray_DESCR((PyArrayObject *)pv))) {
-        PyErr_SetString(PyExc_TypeError,
-                "object array may be self-referencing");
-        Py_DECREF(pv);
+    res = PyObject_CallMethodObjArgs(mod, "oct", o, NULL);
+    Py_DECREF(module);
```
Comment (mattip) [inline numpy/core/src/multiarray/number.c]: "module" is not defined, you probably meant "mod"

File: numpy/core/src/multiarray/number.c [hunk 2]
```diff
@@ -814,213 +814,107 @@ _array_nonzero(PyArrayObject *mp)
...
+    PyObject *res;
+    PyObject *mod = PyImport_ImportModule("builtins");
+    if (mod != NULL) {
         return NULL;
     }
-    pv2 = Py_TYPE(pv)->tp_as_number->nb_long(pv);
-    Py_DECREF(pv);
-    return pv2;
+    res = PyObject_CallMethodObjArgs(mod, "hex", o, NULL);
+    Py_DECREF(module);
```
Comment (mattip) [inline numpy/core/src/multiarray/number.c]: "module" -> "mod" ?

File: numpy/core/src/multiarray/number.c (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/number.c]: Good catch, only tested on python 3 locally

File: numpy/core/tests/test_regression.py [hunk 3]
```diff
@@ -1683,16 +1683,19 @@ def test_object_array_self_reference(self):
...
+            # in python 3, this falls back on operator.index, which fails on
+            # on dtype=object
+            assert_raises(RecursionError, oct, a)
+            assert_raises(RecursionError, hex, a)
 
         # Test the same for a circular reference.
         b = np.array(a, dtype=object)
         a[()] = b
-        assert_raises(TypeError, int, a)
+        assert_raises(RecursionError, int, a)
```
Comment (mattip) [inline numpy/core/tests/test_regression.py]: [RecursionError](https://docs.python.org/3/library/exceptions.html?highlight=recursionerror#RecursionError) is a python3-ism, is there a trick for python 2?

Comment (mattip) [reply numpy/core/tests/test_regression.py]: test_multiarray does:
```
        try:
            Error = RecursionError
        except NameError:
            Error = RuntimeError  # python < 3.5
        assert_raises(Error, bool, self_containing)  # previously stack overflow
```

Comment (eric-wieser) [reply numpy/core/tests/test_regression.py]: Yeah, we handle that elsewhere. Again, good catch!

**Comment by eric-wieser** [comment]
Something fishy going on here on py 2...

**Comment by mattip** [comment]
document in release notes?

**Comment by eric-wieser** [comment]
Not sure it's note-worthy, since this is a bugfix for a bunch of contrived corner-cases. I don't think it has any compatibility ramifications either.

File: numpy/core/src/multiarray/number.c [hunk 4]
```diff
@@ -814,213 +814,112 @@ _array_nonzero(PyArrayObject *mp)
...
+/*
+ * Convert the array to a scalar if allowed, and apply the builtin function
+ * to it. The where argument is passed onto Py_EnterRecursiveCall, in case
+ * there is
+ */
 NPY_NO_EXPORT PyObject *
-array_int(PyArrayObject *v)
+array_scalar_forward(PyArrayObject *v,
+                     PyObject *(*builtin_func)(PyObject *),
+                     const char *where)
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/number.c]: In principle we could use templating instead of function pointers, but we don't really care about speed here.

**Comment by eric-wieser** [comment]
Travis fails with:
```
numpy/core/src/multiarray/number.c: In function ‘array_scalar_forward’:
numpy/core/src/multiarray/number.c:842:9: warning: passing argument 1 of ‘_Py_CheckRecursiveCall’ discards ‘const’ qualifier from pointer target type [enabled by default]
         if (Py_EnterRecursiveCall(where) != 0) {
         ^
In file included from /usr/include/python3.4dm/Python.h:115:0,
                 from numpy/core/src/multiarray/number.c:2:
/usr/include/python3.4dm/ceval.h:80:17: note: expected ‘char *’ but argument is of type ‘const char *’
 PyAPI_FUNC(int) _Py_CheckRecursiveCall(char *where);
                 ^
```

However, this is a python 3.4 bug, not a numpy bug - it's fixed in python/cpython@5fa22fc088ac44c5652107c7c17cda01eaa84a28 (3.4.4)

**Comment by mattip** [comment]
could you either skip the test or do a cast (for specifically that version)?

**Comment by mattip** [comment]
@eric-wieser This pull request is required for PyPy to pass the test suite. There is a regression somewhere from 1.13 wrt handling of string scalars, I am not sure exactly where. Could we mark the pull request for inclusion in 1.14?

**Comment by eric-wieser** [comment]
If this truly is a regression, it should make 1.13.4 - but I'm surprised, as I don't remember changes that would cause this.

**Comment by eric-wieser** [comment]
Milestone added, along with workaround to squash the warning for old python

**Comment by ahaldane** [comment]
Seems like a good cleanup, code LGTM. So it now it allows object arrays (that don't recurse), and uses the python-builtin conversion funcs instead of (badly) trying to do the conversion ourselves.

I don't totally see why this supersedes #10040 - that PR did adjustments of the `tp_` slots. I presume we want to come back to the `tp_` slot issues later. But on it's own this PR looks good.

**Comment by eric-wieser** [comment]
#10040 was needed specifically because we did a bad job of doing the conversion ourselves - our approach didn't special-case string types like the builtin `PyNumber_Long` does.

It does seem a little weird that `str.__int__` does not exist in CPython...

**Comment by mattip** [comment]
@eric-wieser wrt [regression comment above](https://github.com/numpy/numpy/pull/10042#issuecomment-345296038), for PyPy tests started failing after commit 931758e2, "BUG: Make bool(void_scalar) and void_scalar.astype(bool) consistent", which added extra calls to ``PyArray_Scalar`` in ``@from@_to_@to@`` conversions. After this pull request those tests again pass

And yes, it is strange that ``str.__int__`` does not exists but ``int(str)`` works

**Comment by eric-wieser** [comment]
Any idea why that change caused PyPy to fail? Can you point me to the currently failing test on 1.13?

**Comment by mattip** [comment]
That commit AFAICT is not part of a 1.13 release,  as far as ``git tag --contains 931758e2`` and ``git branch --contains 931758e2`` can tell me

These tests newly fail after that commit (since in PyPy the logic in ``PyInt_FromString`` uses ``type.__int__`` which reaches ``gentype_int`` which fails):
- ``numpy.ma.tests.test_mrecords.TestMRecordsImport.test_fromarrays``, raises ``TypeError: cannot convert to a float; scalar object is not a number``
-  `numpy.ma.tests.test_core.TestFillingValues.test_fillvalue_conversion`` raises ``TypeError: cannot convert to an int; scalar object is not a number``
- ``numpy.core.tests.test_item_selection.TestTake.test_simple`` raises ``TypeError: cannot convert to an int; scalar object is not a number''
- ``numpy.core.tests.test_indexing.TestFancyIndexingEquivalence.test_cast_equivalence`` raises ``TypeError: cannot convert to an int; scalar object is not a number``

After this pull request those tests now pass

**Comment by eric-wieser** [comment]
Got it - the problem is that before in conversions we were doing `int(arr.item())`, and after that commit we're doing `int(arr[()])`.

You're right, it's not in 1.13, so this only needs to make it into 1.14 for there to be no regression.

**Comment by ahaldane** [comment]
I think this is ready to go. I'll merge in a little while if there are no further comments.

**Comment by ahaldane** [comment]
Merging. Thanks Eric!

**Pull request merged**